New reflection invoke APIs [draft] (#286)
dotnet/designs: accepted/2022/ReflectionInvoke.md

# Reflection Invoke for 8.0 (draft \ in progress)
#### Steve Harter
#### December 15, 2022

# Background
For additional context, see:
- [Developers using reflection invoke should be able to use ref struct](https://github.com/dotnet/runtime/issues/45152)
- [Originally library issue; needs refreshing\updating](https://github.com/dotnet/runtime/issues/10057)

The invoke APIs and capabilities have essentially remained the same since inception of .NET Framework. Basically, [`MethodBase.Invoke`](https://learn.microsoft.com/dotnet/api/system.reflection.methodbase.invoke):
```cs
public object? Invoke(object? obj, object?[]? parameters);

public object? Invoke(
    object? obj,
    BindingFlags invokeAttr,
    Binder? binder,
    object?[]? parameters,
    CultureInfo? culture);
```

> which is implemented by [MethodInfo](https://learn.microsoft.com/dotnet/api/system.reflection.methodinfo) and [ConstructorInfo](https://learn.microsoft.com/dotnet/api/system.reflection.constructorinfo). Fields do not have an `Invoke()` and instead have `GetValue()` and `SetValue()` methods since there is no invokable code around field access.

Reviewer: Assume that properties are implemented by `MethodInfo` on the accessors?

Author: Yep. Just clarified that.

- [MethodInfo.CreateDelegate()](https://learn.microsoft.com/dotnet/api/system.reflection.methodinfo.createdelegate) which supports a direct method invocation and thus is strongly-typed and not appropriate for loosely-typed invoke scenarios.
- [Dynamic methods](https://learn.microsoft.com/dotnet/framework/reflection-and-codedom/how-to-define-and-execute-dynamic-methods) which are IL-emit based and include a non-trivial implementation.
- Compiled [expression trees](https://learn.microsoft.com/dotnet/csharp/programming-guide/concepts/expression-trees/) which use dynamic method if IL Emit is available but also conveniently fallback to standard reflection when IL Emit is not available.

# 8.0 Goals
The 7.0 release had a 3-4x perf improvement for `Invoke()` by using IL Emit when available and falling back to standard reflection when not available, keeping the same object-based APIs. Although this improvement is significant, it doesn't replace the need to use the IL-Emit based alternatives for the highly-performance-sensitive scenarios such as the `System.Text.Json` serializer. New APIs are required.

For 8.0, there are two primary goals:

> 1) Support byref-like types both for invoking and passing as arguments. An unsafe approach may need to be used for V8. This unblocks scenarios.

Reviewer 2: Using V8 is confusing, could you use common .NET8 instead?

Author: Sure. Thanks

> 2) Support "fast invoke" so that dynamic methods no longer have to be used for performance. Today both `System.Text.Json` and [dependency injection](https://learn.microsoft.com/dotnet/core/extensions/dependency-injection) use IL emit for performance. As a litmus test, these areas will be changed to use the new APIs proposed here.

Reviewer 3: Replacing dynamic methods in `Microsoft.Extensions.DependencyInjection` is more important than replacing dynamic methods in `System.Text.Json`.

Our .NET 8 plan is to keep `Microsoft.Extensions.DependencyInjection` dependency on reflection for trimming and native AOT compatibility. It means that we really need a fast reflection invoke that does not depend on dynamic code generation.

The plan for System.Text.Json is to use source generators for trimming and native AOT compatibility. It means that we do need to get rid of the dynamic code generation in (non-source generated) System.Text.Json to achieve our .NET 8 goals.

cc @eerhardt @davidfowl

Reviewer 4: Right. Today, it's hard to remove reflection from dependency injection because the resolution of instances relies on the knowledge that only exists at runtime (the list of services registered in the container). We could do partial codegen for each module but that could end up being slower (it's similar to partial crossgen).

This is the implementation that gets used at runtime when resolving instances in AOT mode using DI:

https://github.com/dotnet/runtime/blob/5b8ebeabb32f7f4118d0cc8b8db28705b62469ee/src/libraries/Microsoft.Extensions.DependencyInjection/src/ServiceLookup/CallSiteRuntimeResolver.cs

Author: > It means that we really need a fast reflection invoke that does not depend on dynamic code generation and that supports everything that Microsoft.Extensions.DependencyInjection needs.

To clarify: for "fast invoke", the new APIs still require "dynamic code generation" through IL Emit but of course use the "slow invoke" interpreted fallback approach when not available. So DI would use the new reflection APIs instead instead of expressions. An unstated goal is to support https://github.com/dotnet/runtime/issues/66153 which doesn't want to use expressions in Blazor client due to having to bring along the large System.Linq.Expressions.dll.

Reviewer 3: >  for "fast invoke", the new APIs still require "dynamic code generation" through IL Emit but of course use the "slow invoke" interpreted fallback approach when not available. 

For native AOT config, the invocation should be done via a stub generated at build time. This stub is similar to the stub generated via IL Emit and it has similar performance. It is how reflection invoke works in native AOT today. The slow interpreted fallback does not exist in native AOT.

Reviewer 4: it seems like we could delete a large amount of code that uses expression trees in some cases if we could have the CoreCLR version of these APIs use reflection emit under the covers. Then we could just use the built in methods and have them be "best in class" on each of the available platforms.

Reviewer 3: > I'll assume for now "InvokeDirect()` is as fast as possible with perhaps only argument number validated

The set\get parameters have to store the additional type information so that it can maybe validated later. Is that correct? `InvokeDirect` maybe as fast as possible, but the rest is not.

Author: > The set\get parameters have to store the additional type information so that it can maybe validated later. Is that correct?

Not when using the `object` set overloads. The invoker maintains a tuple **(byreference,** **object,** **type)** for each arg and depending on the set method used and whether the target method has byref parameters. For the simple `object`-only case used by DI, for example, we don't need the `byreference` or the `type` for Core at least.

For perf, Core will have both the byref- and object-based emit. NativeAOT may elect to force everything to byrefs to have just one stub.

Author: For DI I'll provide more detail, but given this perf test:
- DI factory logic which is `object`-based with no byref args
- The perf measures calling the factory method. This was earlier obtained from `ActivatorUtilities.CreateFactory(Type, Type[])` but that is not included in the benchmark.
- That factory method was changed to use the new `MethodInvoker`.
- The factory method has 3 parameters, 2 of which are "matched" and 1 is "unmatched" so there's some logic involved in calling the IServiceProvider for the unmatched case.

The breakdown of time spent:
- ~12% for calling the invoke delegate which call the ctor of the target type passing the 3 args.
- ~5% for MethodInvoker.Invoke() overhead
- ~80% for setting the args and preparing to call MethodInvoker.Invoke(). This can vary depending on fixed vs. variable MethodInvoker APIs, but not much. This also includes the DI logic for default values and mapping.

That 80% dwarfs the actual invoke of 12%.

The test ran in ~55ns compared to ~66ns for today's reflection (which uses emit since 7.0). This ~17% gain also includes some other misc optimizations. In summary, this is not getting us much except there is no more alloc for the `object[]` for the args.

Compare this to LINQ with compiled expressions which is ~2.5x faster (~22ns).

Based on this I have a new prototype:
- Added additional Invoke() APIs for fixed-length cases (say up to 8). Added a static MethodInvoker.Invoke(...) or MethodInvoker.GetInvokeDelegate(...) that requires no allocs or setting of values. It will only take `object` args and be focused around existing `MethodBase.Invoke` cases which can only take `object`.
- Perform a manual "AOT" generation of the DI factory lambda code depending how the matching of args occurs. This is based on whether default values are required and the mapping logic. It avoids capturing unnecessary state by the factory's lambda and avoids various logic to check for default values etc.

This new prototype looks promising but will never be as fast as LINQ expressions since that can hard-code the default values and mapping information in the expression. For the happy path, I think I can get it to within 30-50% (so will still be 1.3-1.5x slower, but not 2.5x).

> 

Reviewer 3: I would be useful to get a list of reflection APIs that we believe we would need to eliminate dynamic codegen from Microsoft.Extensions.DependencyInjection and System.Text.Json without incurring unacceptable perf regressions (two lists - one for each library).

Reviewer 4: - [Array.CreateInstance](https://github.com/dotnet/runtime/blob/5b8ebeabb32f7f4118d0cc8b8db28705b62469ee/src/libraries/Microsoft.Extensions.DependencyInjection/src/ServiceLookup/CallSiteRuntimeResolver.cs#L165)
- [ConstructorInfo.Invoke](https://github.com/dotnet/runtime/blob/5b8ebeabb32f7f4118d0cc8b8db28705b62469ee/src/libraries/Microsoft.Extensions.DependencyInjection/src/ServiceLookup/CallSiteRuntimeResolver.cs#L69)

Some of the overhead will be in DI itself, but these are the 2 methods that are "reflection invoke based" in the hot path. The other uses of reflection are metadata based (ConstructorInfo, ParamterInfo etc), but I'm assuming we're not talking about those calls.

Author: I added the sections for DI\STJ listing their usage of reflection.

See also [C++\CLI documentation](https://learn.microsoft.com/en-us/cpp/extensions/interior-ptr-cpp-cli?view=msvc-170).

An interior pointer is created in several ways:
- [System.TypedReference](https://learn.microsoft.com/en-us/dotnet/api/system.typedreference). Since this a byref-like type, it is only stack-allocated. Internally, it uses the `ref byte` approach below.
- Using `ref byte`. This is possible in 7.0 due to the new "ref field" support. Previously, there was an internal `ByReference<T>` class that was used. This `ref byte` approach is used today in reflection invoke when there are <=4 parameters.
- Using `IntPtr` with tracking or pinning. Currently, tracking is only supported internally through the use of a "RegisterForGCReporting()" mechanism. This approach is used today in reflection invoke when there are >=5 parameters.

This design proposes the use of `TypedReference` for the safe version of APIs, and either `ref byte` or `IntPtr` approaches for the unsafe version. The `ref byte` and `IntPtr` approaches require the use of pointers, through the internal `object.GetRawData()` and other means.

## TypedReference

> The internal reflection implementation will not likely be based on `TypedReference` -- instead it will take, from public APIs, `TypedReference` instances and "peel off" the interior pointer as an `IntPtr` along with tracking to support GC. The use of `TypedReference` is essentially to not require the use of unsafe code in the public APIs. Unsafe pubic APIs will likely be added as well that should have slightly better performance characteristics.

Reviewer 5: ```suggestion
The internal reflection implementation will not likely be based on `TypedReference` -- instead it will take, from public APIs, `TypedReference` instances and "peel off" the interior pointer as an `IntPtr` along with tracking to support GC. The use of `TypedReference` is essentially to not require the use of unsafe code in the public APIs. Unsafe public APIs will likely be added as well that should have slightly better performance characteristics.
```

### Existing usages

> `TypedReference` is used in `FieldInfo` via [`SetValueDirect(TypeReference obj, object? value)`](https://learn.microsoft.com/dotnet/api/system.reflection.fieldinfo.setvaluedirect) and [`object? GetValueDirect(TypeReference obj)`](https://learn.microsoft.com/dotnet/api/system.reflection.fieldinfo.getvaluedirect). The approach taken by `FieldInfo` will be expanded upon in the design here. The existing approach supports the ability to get\set a field on a target value type without boxing. Boxing the target through through [`FieldInfo.SetValue()`](https://learn.microsoft.com/dotnet/api/system.reflection.fieldinfo.setvalue) instead is useless since the changes made to the boxed value type would not be reflected back to the original value type instance.

Reviewer 6: ```suggestion
`TypedReference` is used in `FieldInfo` via [`SetValueDirect(TypeReference obj, object? value)`](https://learn.microsoft.com/dotnet/api/system.reflection.fieldinfo.setvaluedirect) and [`object? GetValueDirect(TypeReference obj)`](https://learn.microsoft.com/dotnet/api/system.reflection.fieldinfo.getvaluedirect). The approach taken by `FieldInfo` will be expanded upon in the design here. The existing approach supports the ability to get\set a field on a target value type without boxing. Boxing the target through [`FieldInfo.SetValue()`](https://learn.microsoft.com/dotnet/api/system.reflection.fieldinfo.setvalue) instead is useless since the changes made to the boxed value type would not be reflected back to the original value type instance.
```

> `TypedReference` is also used by the undocumented `__arglist` along with `System.ArgIterator`. The approach taken by `__arglist` will not be leveraged or expanded upon in this design. It would, however, allow a pseudo-strongly-typed approach like

Reviewer: I thought `__arglist` was Windows only. Is that not the case or do we plan on extending it?

Author: Yes that's the case -- Windows only and no plans to change that for V8 although it has been requested.

I was just pointing out here that Roslyn and the runtime already support `TypedReference` here.

TypedReference tr1 = __makeref(i); // okay

Span<int> span = stackalloc int[42];
TypedReference tr2 = __makeref(span); // Error CS1601: Cannot make reference to variable of type 'Span<int>'
```

An 8.0 ask from Roslyn is to allow this to compile. See [C# ask for supporting TypedReference + byref-like types](https://github.com/dotnet/roslyn/issues/65255).

Since `TypedReference` internally stores a reference to the **storage location**, and not the actual value or managed-reference-passed-by-value, it effectively supports the `ref\out\in` modifiers without using the `ref` keyword. Attempting to use the `ref` keyword is not allowed:
```cs
int i = 42;
TypedReference tr = __makeref(ref i); // error CS1525: Invalid expression term 'ref'
```
and not necessary due to `__refvalue`:
```cs
int i = 42;
TypedReference tr = __makeref(i);
__refvalue(tr, int) = 100;
Console.WriteLine(i); // 100
```
or by using `ref` along with `__refvalue`:
```cs
int i = 42;
TypedReference tr = __makeref(i);
ChangeIt(tr);
Console.WriteLine(i); //100

public static void ChangeIt(TypedReference tr)
{
    ref int i = ref __refvalue(tr, int);
    i = 100;
}
```

### Byref-like support + simplification constraints

> Below are some simplfication constraints that may help with any Roslyn implementation around lifetime rules.

Reviewer 3: ```suggestion
Below are some simplification constraints that may help with any Roslyn implementation around lifetime rules.
```

}

public ref struct MyRefStruct
{
    public Span<int> _span;
}
```

If allowed today in strongly-typed manner, then reflection should also allow:
```cs
internal class Program
{
    static void Main()
    {

        Span<int> span = stackalloc int[42];
        TypedReference tr = __makeref(span); // Assume we allow instead of CS1601

        // Using a proposed invoke API; calling should be supported passing byvalue
        MethodInfo mi1 = typeof(Program).GetMethod(nameof(ChangeIt1));
        mi1.InvokeDirect(target: default, arg1: tr);

        // and supported passing byref
        MethodInfo mi2 = typeof(Program).GetMethod(nameof(ChangeIt2));
        mi2.InvokeDirect(target: default, arg1: tr);

        // Just like these methods can be called today:
        ChangeIt1(span);
        ChangeIt2(ref span);
    }

    public static void ChangeIt1(Span<int> span) { }
    public static void ChangeIt2(ref Span<int> span) { }
}
```

FWIW `__arglist` currently compiles with byref-like types:
```cs
Span<byte> s1 = stackalloc byte[1];
Span<byte> s2 = stackalloc byte[11];
Span<byte> s3 = stackalloc byte[111];
CallMe(__arglist(s1, s2, s3));

static unsafe void CallMe(__arglist)
{

>     // However, when enumerating __arglist here, the runtime throws when accessing a byref-like

Reviewer 6: Is the runtime tracking lifetime properly in this case?

Reviewer: FWIW the compiler does do lifetime tracking here.

    // type although that limitation could be removed (just a runtime limitation; not compiler)
```

## Variable-length, safe collections
(todo; see https://github.com/dotnet/runtime/issues/75349. _I have a local prototype that does a callback that doesn't require any new language or runtime features however it is not a clean programming model. The callback is necessary to interop with hide our internal GC "tracking" feature via "RegisterForGCReporting()"._)

## Exception handling
Todo; discuss using `BindingFlags.DoNotWrapExceptions` semantics only (no wrapping of exceptions).

# Proposed APIs
(work in progress; various prototypes exist here)
## TypedReference
```diff
namespace System
{
    public ref struct TypedReference
    {

>         // Equivalent of __makeref (except for a byref-like type since they can't be a generic parameter)

Reviewer 6: Reference https://github.com/dotnet/runtime/issues/65112

Author: Thanks. Added

> +       public static TypedReference Make<T>(ref T? value);

Reviewer: This is strictly less powerful than `__makeref` though because it can't support `ref struct`. Until we get support for `ref struct` generic parameters my recommendation would be to use `__makeref` everywhere.

Reviewer 6: > my recommendation would be to use __makeref everywhere.

Would we then advocate and evangelize it usage? I thought it was considered bad form to use any `__` keywords nowadays.

Reviewer: One of the explicit goals of this work is to enable `ref struct` to participate here. Rather than say use `TypedReference.Make<T>` for most types but `__makeref` for `ref struct` I prefer the simpler, context-free, advice of use `__makeref`. 

I agree that `__` is explicitly unsupported now but part of this work is basically us defining these operators, adding docs and telling people when we support them.

Reviewer 7: Based on separate discussion, my feeling here is: let's use `__makeref` etc. for the time being, but eventually make sure we introduce the ability to express `TypedReference.Make<T>` in the language in subsequent releases, and steer people towards that once it's available.

        // Helper used for boxed or loosely-typed cases
+       public static TypedReference Make(ref object value, Type type);

        // Equivalent of __refvalue
+       public ref T GetValue<T>();
        // Used for byref-like types or loosely-typed cases with >4 parameters

> +       public readonly unsafe ref byte TargetRef { get; }

Reviewer 6: This is for an on stack array, right?

Author: Yes it would be used to help "peel off" the reference so it can be passed as a non-Typed referenced. See this [prototype ](https://github.com/dotnet/runtime/compare/main...steveharter:runtime:InvokeWithStj#diff-a6a6bf9694f968751e223700b17bb8588bddc4a9ed62e0a9d8534bf390e60891R68) for the simple case of a single parameter (property value).

    }
}
```

## MethodInfo
(these may be extension methods instead)

```diff
namespace System.Reflection
{
    public abstract class MethodBase
    {
        // Helpers for <= 4 parameters

        // Note that 'default(TypedReference)' can be specified for any "not used" arguments, such as for
        // static methods ('target' is 'default'), 'void'-returning methods ('result' is 'default') or
        // for any trailing unused arguments (e.g. 'arg4' can be 'default').

        // We may also want to consider using 'ref byte' instead of or in addition to 'TypedReference'
        // along with new helper methods like 'public static ref byte GetRawData(object o)'
+       [System.CLSCompliantAttribute(false)]
+       public virtual void InvokeDirect(

> +           TypedReference target,
> +           TypedReference result);

Reviewer 8: Since this is a bigger struct, it's going to be passed by reference on many platforms anyway. Should we make it explicit?

```suggestion
+           in TypedReference target,
+           out TypedReference result);
```

Reviewer 3: `TypedReference` is same size as `Span<T>` (2*IntPtr.Size). We do not pass `Span<T>` via `in` and it would be de-optimization in typical case. Also, the .NET Framework Design Guidelines are advising against `in` in public APIs (look for "DO NOT pass value types by read-only reference (in).").

> `out TypedReference result`

This would have to be `in TypedReference result`. The `TypedReference` is meant to point to a location to store the result.

Author: FWIW Roslyn actually doesn't allow `TypedReference` to be passed today via `in\ref\out` -- CS1601. I don't think we need to change this. Also that would only work if bound to a ref field in a ref struct since it is be stack-allocated.

+       [System.CLSCompliantAttribute(false)]
+       public virtual void InvokeDirect(
+           TypedReference target,
+           TypedReference arg1,

> +           TypedReference result);

Reviewer 2: What will be TypedReference value of method return a void?

Author: I'm thinking just `default(TypedReference)` would be used. I believe it mentions this just above.

# .NET 8 Goals
The .NET 7 release had a 3-4x perf improvement for the existing `Invoke()` APIs by using IL Emit when available and falling back to standard reflection when IL Emit is not available. Although this improvement is significant, it still doesn't replace the need to use the IL-Emit based alternatives (dynamic methods and expression trees) for highly-performance-sensitive scenarios including the `System.Text.Json` serializer. New APIs are required that don't have the overhead of the existing `Invoke()` APIs.

For .NET 8, there are two primary goals:
1) Support byref-like types both for invoking and passing as arguments; this unblocks various scenarios. An unsafe approach may used for .NET 8 if support for `TypedReference` isn't addressed by Roslyn (covered later).
2) Support "fast invoke" so that dynamic methods no longer have to be used for performance. Today both `STJ (System.Text.Json)` and [DI (dependency injection)](https://learn.microsoft.com/dotnet/core/extensions/dependency-injection) use IL Emit for performance although DI uses emit through expressions while STJ uses IL Emit directly.

## STJ and DI

> As a litmus test, STJ and DI will be changed to use the new APIs proposed here. This is more important to DI since, unlike STJ which has a source generator that can avoid reflection, DI must continue to use reflection. See also https://github.com/dotnet/runtime/issues/66153 which should be addressed by having a fast constructor invoke that can be used by DI.

Author: That is the thinking - if it's not fast enough for internal use, it may not be fast enough for others. However, I have a prototype for STJ using the proposed APIs and has a 5% regression in deserialization (note that without any Emit, deserialization is ~2x slower, so getting it to just 5% is pretty significant). That was the reason I also added the `PropertyInfo.CreateGetter\SetterDelegate` to match how STJ uses delegates today. There may also be some gains in going down an unsafe approach, but I haven't measured the perf of that yet.

For DI, I hope to use the new APIs directly.

### Existing usages
`TypedReference` is used in `FieldInfo` via [`SetValueDirect(TypeReference obj, object? value)`](https://learn.microsoft.com/dotnet/api/system.reflection.fieldinfo.setvaluedirect) and [`object? GetValueDirect(TypeReference obj)`](https://learn.microsoft.com/dotnet/api/system.reflection.fieldinfo.getvaluedirect). The approach taken by `FieldInfo` will be expanded upon in the design here. The existing approach supports the ability to get\set a field on a target value type without boxing. Boxing the target through [`FieldInfo.SetValue()`](https://learn.microsoft.com/dotnet/api/system.reflection.fieldinfo.setvalue) is useless since the changes made to the boxed value type would not be reflected back to the original value type instance.

> `TypedReference` is also used by the undocumented `__arglist` along with `System.ArgIterator` although `__arglist` is Windows-only. The approach taken by `__arglist` will not be leveraged or expanded upon in this design. It would, however, allow a pseudo-strongly-typed approach like

Reviewer: Should we add a top level item to track expanding `__arglist` to non-windows platforms? Most of other non-trivial work items have a top level item and my guess is this is non-trivial work.

Reviewer 6: The idea of native varargs is tracked at https://github.com/dotnet/runtime/issues/48796.

Reviewer: We should link to that here then.

+           TypedReference arg2,
+           TypedReference result);

+       [System.CLSCompliantAttribute(false)]
+       public virtual void InvokeDirect(
+           TypedReference target,
+           TypedReference arg1,
+           TypedReference arg2,
+           TypedReference arg3,
+           TypedReference result);

+       [System.CLSCompliantAttribute(false)]
+       public virtual void InvokeDirect(
+           TypedReference target,
+           TypedReference arg1,
+           TypedReference arg2,
+           TypedReference arg3,
+           TypedReference arg4,
+           TypedReference result);

        // Unsafe (todo: more on this; samples)
+       public virtual unsafe void InvokeDirect(
+           TypedReference target,
+           TypedReference* parameters,
+           TypedReference result);

       // Also consider a safe variable-length callback that does tracking internally
       // since we don't support a safe variable-length collection mechanism
       // or the ability to use a Span<TypedReference>
       // (todo: more on this?; prototype exists)
    }
}
```

## PropertyInfo \ FieldInfo
```diff
namespace System.Reflection
{

>     public abstract class PropertyInfo

Reviewer 4: We should make sure these APIs allow ASP.NET to delete this:

https://github.com/dotnet/aspnetcore/blob/4535ea1263e9a24ca8d37b7266797fe1563b8b12/src/Shared/PropertyHelper/PropertyHelper.cs

    {
+       [System.CLSCompliantAttribute(false)]
        public virtual void GetValueDirect(TypedReference target, TypedReference result);

+       [System.CLSCompliantAttribute(false)]
        public virtual void SetValueDirect(TypedReference target, TypedReference value);

        // Possible for performance in System.Text.Json:
+       public virtual Func<object, TValue> CreateGetterDelegate<TValue>();

> +       public virtual Action<object, TValue> CreateSetterDelegate<TValue>();

Reviewer 4: https://github.com/dotnet/aspnetcore/blob/4535ea1263e9a24ca8d37b7266797fe1563b8b12/src/Shared/PropertyHelper/PropertyHelper.cs#L73-L92

Author: The proposed TypedReference APIs will be able to take a boxed `object` and use the internal reference, so object is supported that way.

However, if there are a variable number of parameters that need to be passed, you'll still need to pass them down somehow (e.g. `object[]` or an unsafe approach (e.g. `TypedReference*[] \ void*`) until an Invoke can be done.

# Reflection Invoke for 8.0 (draft \ in progress)
#### Steve Harter
#### April 13, 2023

# Background
For additional context, see:
- [Developers using reflection invoke should be able to use ref struct](https://github.com/dotnet/runtime/issues/45152)

> - [Originally library issue; needs refreshing\updating](https://github.com/dotnet/runtime/issues/10057)

Reviewer 9: Good read overall. Though I found the use of "\\" instead of "/" jarring
```suggestion
- [Originally library issue; needs refreshing/updating](https://github.com/dotnet/runtime/issues/10057)
```

Ditto other places.

- Compiled [expression trees](https://learn.microsoft.com/dotnet/csharp/programming-guide/concepts/expression-trees/) which use dynamic methods if IL Emit is available but also conveniently falls back to standard reflection when not available. Using an expression to invoke a member isn't intuitive and brings along the large `System.Linq.Expressions.dll` assembly.

# .NET 8 Goals
The .NET 7 release had a 3-4x perf improvement for the existing `Invoke()` APIs by using IL Emit when available and falling back to standard reflection when IL Emit is not available. Although this improvement is significant, it still doesn't replace the need to use the IL-Emit based alternatives (dynamic methods and expression trees) for highly-performance-sensitive scenarios including the `System.Text.Json` serializer. New APIs are required that don't have the overhead of the existing `Invoke()` APIs.

For .NET 8, there are two primary goals:

> 1) Support byref-like types both for invoking and passing as arguments; this unblocks various scenarios. An unsafe approach may used for .NET 8 if support for `TypedReference` isn't addressed by Roslyn (covered later).

Reviewer 3: I believe that we know by know that we are not going to get safe TypedReferences from Roslyn in .NET 8.

Author: Sure - I'll clarify that and move to the "future" section.

Reviewer: Think there is a simple design where we have safe APIs for constructing `TypedReference` instances and getting the values back from them. The only type `unsafe` would come into the mix is when you attempted to extract the value of a `ref struct` by `ref`. That would put some burden on the implementation here to be careful with the values but consumers shouldn't need to. 

The conversation on whether or not this will work for the reflection APIs keeps lagging though so it's hard to tell if we should be investing in it.

2) Support "fast invoke" so that using IL Emit with dynamic methods has little to no performance advantage. Today both `STJ (System.Text.Json)` and [DI (dependency injection)](https://learn.microsoft.com/dotnet/core/extensions/dependency-injection) use IL Emit for performance although DI uses emit through expressions while STJ uses IL Emit directly.

>     - Although the new APIs will be zero alloc (no `object` and boxing required) and performace faster than standard reflection, it will not necessarily be as fast as hand-coded IL Emit for specific scenarios that can optimize for any constraints such as fixed-size list of parameters or by not doing full defaulting and validation of values. Note that property get\set is a subset of this case since there is either a return value (for _get_) or a single parameter for _set_ and since property get\set is such a common case with serialization, this design does propose a separate API for this common case to enable maximum performance.

Reviewer 10: ```suggestion
    - Although the new APIs will be zero alloc (no `object` and boxing required) and performance faster than standard reflection, it will not necessarily be as fast as hand-coded IL Emit for specific scenarios that can optimize for any constraints such as fixed-size list of parameters or by not doing full defaulting and validation of values. Note that property get\set is a subset of this case since there is either a return value (for _get_) or a single parameter for _set_ and since property get\set is such a common case with serialization, this design does propose a separate API for this common case to enable maximum performance.
```

# Design with managed pointers
In order to address the limitations and issues, a least-common-denominator approach of using _managed pointers_ for the parameters, target and return value. This supports `in\ref\out` aliasing and the various classifications of types including value types, reference types, pointer types and byref-like types. Essentially this is a different way to achieve "boxing" in order to have a single representation of any parameter.

> However, since managed pointers require a reference to a storage location, they does not directly support the same loosely-coupled scenarios as using `object` + boxing for value types. The proposed APIs, however, do make interacting with `object` possible with the new API for the cases that do not require `in\ref\out` variable aliasing for example.

Reviewer 11: ```suggestion
However, since managed pointers require a reference to a storage location, they do not directly support the same loosely-coupled scenarios as using `object` + boxing for value types. The proposed APIs, however, do make interacting with `object` possible with the new API for the cases that do not require `in\ref\out` variable aliasing for example.
```

Today, a managed pointer is obtained safely through the `ref` keyword in C#. It references a storage location of an object or value type which can be a stack variable, static variable, parameter, field or array element. If the storage location is a field or array element, the managed pointer is referred to as an "interior pointer" which is supported by GC meaning that GC won't collect the owning object even if there are only interior pointers to it.

> A managed pointer can becreated in several ways:
> - [System.TypedReference](https://learn.microsoft.com/en-us/dotnet/api/system.typedreference). Since this a byref-like type, it is only stack-allocated. Internally, it uses a `ref byte` approach along with a reference to a `Type`. Note `TypeReference` is a special type and has its own opcodes (mkrefany, refanytype, refanyval) which translate to C# keyworkds (`__makeref`, `__reftype`, `__refvalue`).

Reviewer 10: ```suggestion
A managed pointer can be created in several ways:
- [System.TypedReference](https://learn.microsoft.com/en-us/dotnet/api/system.typedreference). Since this a byref-like type, it is only stack-allocated. Internally, it uses a `ref byte` approach along with a reference to a `Type`. Note `TypeReference` is a special type and has its own opcodes (mkrefany, refanytype, refanyval) which translate to C# keywords (`__makeref`, `__reftype`, `__refvalue`).
```

- Using `ref <T>` for the strongly typed case or `ref byte` for the loosely-typed case. This is expanded in 7.0 due to the new "ref field" support. Previously, there was an internal `ByReference<T>` class that was used and in 7.0 this was changed to `ByReference` in 8.0 which no longer maintains the `<T>` type and internally just contains `ref byte`. This `ByReference` type is used today in reflection invoke when there are <=4 parameters.
- Using unsafe `void*` (or `IntPtr`) with GC tracking or pinning to make the use GC-safe. Tracking is supported internally through the use of a newer "RegisterForGCReporting()" mechanism. This approach is used today in reflection invoke when there are >=5 parameters.

# Proposed APIs

## MethodInvoker
This ref struct is the mechanism to specify the target + arguments (including return value) and supports these mechanisms:

> - `object` (including boxing). Supports loose coupling scenarios are supported, like reflection today.

Reviewer 11: ```suggestion
- `object` (including boxing). Loose coupling scenarios are supported, like reflection today.
```

- `ref <T>`. Supports new scenarios as mentioned earlier; type must be known ahead-of-time and due to no language support, cannot be a byref-like type like `Span<T>`.
- `void*`. Unsafe cases used to support byref-like types in an unsafe manner.
- `TypedReference`. Optional for now; pending language asks, it may make supporting byref-like types a safe operation.

```cs
namespace System.Reflection
{

>      public ref struct MethodInvoker

Reviewer: Is this actually a mutable `struct` or was `readonly` omitted from the doc by mistake?

Author: Yes it is mutable.

     {
        // Zero-arg case:
        public MethodInvoker()

        // Variable-length number of arguments:

>         public unsafe MethodInvoker(ArgumentValue* argumentStorage, int argCount)

Reviewer 10: Would it be possible to use a `Span<ArgumentValue>` instead?
```csharp
public MethodInvoker(Span<ArgumentValue> values)
```

Author: Perhaps; for now I moved this to the "future" section since it is somewhat speculative based on perf.

        // Fixed length (say up to 8)
        public MethodInvoker(ref ArgumentValuesFixed values)

        // Dispose needs to be called with variable-length case

>         public void Dispose()

Reviewer 10: Why is `Dispose()` needed for the variable-length case but not the fixed-length case?

Author: The variable-length case hooks up in a low level to the runtime to track these references to support the GC and needs to be unregistered once finished.

>         public void Dispose()

Reviewer 11: I'm wondering if having Dispose which is only required in some cases won't trigger false positives in disposable analyzers.

Author: Good question.

        // Target
        public object? GetTarget()

>         public ref T GetTarget<T>()

Reviewer: Did you consider a `GetTarget` method that can return the `TypedReference`? The current design has a lack of symmetry with the `SetTarget` methods.

Author: Yes it should be added; currently that is a CS1599 (can't return a TypedReference) and I didn't call it out.

However, I'm pushing these TypedReference cases to the "future" section for now.

        public void SetTarget(object value)
        public void SetTarget(TypedReference value)
        public unsafe void SetTarget(void* value, Type type)
        public void SetTarget<T>(ref T value)

        // Arguments
        public object? GetArgument(int index)
        public ref T GetArgument<T>(int index)

        public void SetArgument(int index, object? value)
        public void SetArgument(int index, TypedReference value)
        public unsafe void SetArgument(int index, void* value, Type type)
        public void SetArgument<T>(int index, ref T value)

        // Return
        public object? GetReturn()
        public ref T GetReturn<T>()

        public void SetReturn(object value)
        public void SetReturn(TypedReference value)
        public unsafe void SetReturn(void* value, Type type)
        public void SetReturn<T>(ref T value)

>         // Invoke direct (limited validation and defaulting)

Reviewer 3: What does limited validation mean precisely?

I think it would be useful to have fully unsafe API that does not do any validation and just assumes that you have passed in arguments of the right types. Safety and performance wise, such API would be closest to Reflection.Emit where we do not perform any validation either.

Author: What I was thinking was to at least verify values type are the correct type to prevent destabilizing the system in a nondeterministic manner.  This can be done most efficiently in IL with castclass with minimal impact on perf. Verification on reference types should also happen, but I believe this comes for free at least.

However, the exceptions thrown will be CastClassException I believe, not a nice "ArgumentException" unless we do extra work ahead of time to throw ArgumentException or do extra work with a try\catch and back-walk the validation if CastClassException is thrown.

I was also planning on doing "defaulting" for `out` values for value types if no default values were provided by the caller. This is especially common for return values. Unfortunately, it is not easy\fast to see check if a managed reference for an argument is `in` vs `ref` vs `out` so perhaps that should be deferred for now except for the return value which is always `out`.

Reviewer 3: I think it would be useful to have a completely unsafe API as one of the options. The completely unsafe API would not do any argument validation. It would be closest to what one can get with RefEmit where the user is responsible for emitting the right IL and where mistake can destabilize the system too.

> Verification on reference types should also happen, but I believe this comes for free at least.

Validation of reference types is not free. It is more expensive than validation of value types.

Reviewer 4: I agree, we have the unsafe API as well.

Author: Updated: the `InvokeDirectUnsafe()` APIs won't do any validation.

>         public unsafe void InvokeDirect(MethodBase method)

Reviewer 3: Nit: The unsafe API with non-validation should have `Unsafe` in the name.

Author: In most other areas we don't add "Unsafe" since it is a bit redundant since it's in the signature and requires unsafe context to call. I assume then "Unsafe" makes it clearer that this case is more dangerous than the others.

        // Invoke (same validation and defaulting as reflection today)
        public void Invoke(MethodBase method)
     }

     // This is used to define the correct storage requirements for the MethodInvoker variable-length cases.
     // Internally it is based on 3 IntPtrs that are for 'ref', 'object value' and 'Type':
     // - 'ref' points to either the 'value' location or a user-provided location with "void*", "ref <T>" or TypedReference.
     // - 'object value' captures any user-provided object value in a GC-safe manner.
     // - 'Type' is used in "void*", "ref <T>" and TypedReference cases for validation and in rare cases to prevent
     //    Types from being GC'd.
     public struct ArgumentValue { }
```

## ArgumentValuesFixed

> This class is used for cases where the known arguments are small.

Reviewer: ```suggestion
This struct is used for cases where the known arguments are small.
```

```cs
namespace System.Reflection
{

>     public ref partial struct ArgumentValuesFixed

Reviewer: Why are you using this struct vs. a `Span<object>` which is backed by an `[InlineArray(8)]` struct that you slice to the appropriate length?

Author: That struct actually has 3 different types of storage (object, ByReference, and Type) and they need to be grouped something like this in order to register efficiently with the GC tracking support:
```cs
public unsafe ref struct ArgumentValuesFixed
  {
      internal object? _obj0;
      internal object? _obj1-to-lengthl

      internal RuntimeType? _type0;
      internal RuntimeType? _type1-to-lengthl

      internal ref byte _ref0;
      internal ref byte _ref1-to-lengthl
}
```

Also, I moved this type to the "future" section and will add it back, if necessary, based on perf.

        public ArgumentValuesFixed(object? obj1, object? o2, object? o3, object? o4)
        public ArgumentValuesFixed(object? obj1, object? o2, object? o3, object? o4, object? o5)
        public ArgumentValuesFixed(object? obj1, object? o2, object? o3, object? o4, object? o5, object? o6)
        public ArgumentValuesFixed(object? obj1, object? o2, object? o3, object? o4, object? o5, object? o6, object? o7)
        public ArgumentValuesFixed(object? obj1, object? o2, object? o3, object? o4, object? o5, object? o6, object? o7, object? o8)
    }
}
```

## TypedReference
This is currently optional and being discussed. If `TypedReference` ends up supporting references to byref-like types like `Span<T>` then it will be much more useful otherwise just the existing `ref <T>` API can be used instead. The advantage of `TypedReference` is that it does not require generics so it can be made to work with `Span<T>` easier than adding a feature that would allowing generic parameters to be a byref-like type.

To avoid the use of C#-only "undocumented" keywords, wrappers for `__makeref`, `__reftype`, `__refvalue` which also enable other languages.

> ```diff

Author: Because TypedReference is an existing class with existing members.

+       public static TypedReference Make<T>(ref T? value);
+       public static unsafe TypedReference Make(Type type, void* value);
        // Helper used for boxed or loosely-typed cases
+       public static TypedReference Make(ref object value, Type type);

        // Equivalent of __refvalue
+       public ref T GetValue<T>();

        // Equivalent of __reftype
+       public Type Type { get; };
    }
}
```

## PropertyInfo \ FieldInfo
For `PropertyInfo`, this is an alternative of using more heavy-weight `MethodInvoker`. For `FieldInfo`, this expands on the existing `Set\GetValueDirect` to also use `TypedReference` for the `value`.

```diff
namespace System.Reflection
{
    public abstract class PropertyInfo
    {
+       [System.CLSCompliantAttribute(false)]
        public virtual void GetValueDirect(TypedReference obj, TypedReference result);

+       [System.CLSCompliantAttribute(false)]
        public virtual void SetValueDirect(TypedReference obj, TypedReference value);

        // Possible for performance in System.Text.Json:
+       public virtual Func<object, TValue> CreateGetterDelegate<TValue>();

> +       public virtual Action<object, TValue> CreateSetterDelegate<TValue>();

Reviewer 10: Perhaps `CreateGetValueDelegate` and `CreateSetValueDelegate`, to match `GetValue`, `GetValueDirect`, etc.

Author: Thanks. Make sense.

    }

    public abstract class FieldInfo
    {
+       [System.CLSCompliantAttribute(false)]
        public virtual void GetValueDirect(TypedReference obj, TypedReference result);

+       [System.CLSCompliantAttribute(false)]
        public virtual void SetValueDirect(TypedReference obj, TypedReference value);

        // Possible for performance in System.Text.Json:
+       public virtual Func<object, TValue> CreateGetterDelegate<TValue>();

> +       public virtual Action<object, TValue> CreateSetterDelegate<TValue>();

Reviewer 10: Perhaps `CreateGetValueDelegate` and `CreateSetValueDelegate`, to match `GetValue`, `GetValueDirect`, etc.

Author: (dup of above)

    }
}
```

## Examples
### Fixed-length arguments
```cs
MethodInfo method = ... // Some method to call
ArgumentValuesFixed values = new(4); // 4 parameters

> InvokeContext context = new InvokeContext(ref values);

Reviewer 10: Should `InvokeContext` be `MethodInvoker` in these examples?

Author: Yes thanks.

context.SetArgument(0, new MyClass());
context.SetArgument(1, null);
context.SetArgument(2, 42);
context.SetArgument(3, "Hello");

// Can inspect before or after invoke:
object o0 = context.GetArgument(0);
object o1 = context.GetArgument(1);
object o2 = context.GetArgument(2);
object o3 = context.GetArgument(3);

context.InvokeDirect(method);
int ret = (int)context.GetReturn();
```

### Fixed-length object arguments (faster)
```cs
ArgumentValuesFixed args = new(new MyClass(), null, 42, "Hello");
InvokeContext context = new InvokeContext(ref args);
context.InvokeDirect(method);
```

### Variable-length object arguments
Unsafe and slightly slower than fixed-length plus requires `using` or `try\finally\Dispose()`.
```cs
unsafe
{
    ArgumentValue* args = stackalloc ArgumentValue[4];

>     using (InvokeContext context = new InvokeContext(ref args))

Reviewer 10: Should this be `= new MethodInvoker(args, 4)` instead?

    {
        context.SetArgument(0, new MyClass());
        context.SetArgument(1, null);
        context.SetArgument(2, 42);
        context.SetArgument(3, "Hello");
        context.InvokeDirect(method);
    }
}
```

### Avoiding boxing
Value types can be references to avoid boxing.

```cs
int i = 42;
int ret = 0;
ArgumentValuesFixed args = new(4);
InvokeContext context = new InvokeContext(ref args);
context.SetArgument(0, new MyClass());
context.SetArgument(1, null);
context.SetArgument<int>(2, ref i); // No boxing (argument not required to be byref)
context.SetArgument(3, "Hello");
context.SetReturn<int>(ref ret); // No boxing; 'ret' variable updated automatically
context.InvokeDirect(method);
```

### Pass a `Span<T>` to a method
```cs
Span<int> span = new int[] { 42, 43 };
ArgumentValuesFixed args = new(1);

unsafe
{
    InvokeContext context = new InvokeContext(ref args);
#pragma warning disable CS8500    
    void* ptr = (void*)new IntPtr(&span);
#pragma warning restore CS8500    
    // Ideally we can use __makeref(span) instead of the above.

    context.SetArgument(0, ptr, typeof(Span<int>));
    context.InvokeDirect(method);
}
```

> # Design ext

Reviewer 10: "extension"?

Author: Will change to "addendum"; just placing less-important details and findings in an area that isn't up-front and cluttering.

**A `TypedReference` can't reference another `TypedReference`**
```cs
TypedReference tr = ...
TypedReference tr2 = __makeref(tr); //  error CS1601: Cannot make reference to variable of type 'TypedReference'
```

This is similar to trying to make a `TypedReference` to a `Span<int>` above -- both are byref-like types.

However, this limitation for `TypedReference-to-TypedReference` does **not** need to be removed for the goals in this design if it helps with simplifying lifetime rules.

**Just support what is allowed today.** No new capabilities are expected; reflection invokes existing methods which must have already been compiled according to existing rules:
```cs
internal class Program
{
    static void Main()
    {
        // Case 1
        MyRefStruct rs = default;
        TypedReference tr = __makeref(rs._span); // Assume we allow instead of CS1601
        ChangeIt(tr);

        // Case 2
        Span<int> heap = new int[42];
        Span<int> stack = stackalloc int[42];
        CallMe(ref stack, heap); // okay
        CallMe(ref heap, stack); // CS8350 as expected
    }

    public static void ChangeIt(TypedReference tr)
    {
        // This compiles today:

>         ref Span<int> s = ref __refvalue(tr, Span<int>);

Reviewer: I think the easiest path forward for `TypedReference` safely holding a `Span<int>` is for this to turn into an `unsafe` operation. This is where the danger really sets in, unless we significantly complicate the existing lifetime rules, hence this is where `unsafe` should go. Essentially this is when the rules break down if we allow capturing `Span<int>` safely. 

Note: specifically I mean when `ref __refvalue(expr, T)` and `T` is a `ref struct`.

        // Using a proposed invoke API; calling should be supported passing byvalue
        MethodInfo mi1 = typeof(Program).GetMethod(nameof(ChangeIt1));
        mi1.InvokeDirect(target: default, arg1: tr);

        // and supported passing byref
        MethodInfo mi2 = typeof(Program).GetMethod(nameof(ChangeIt2));
        mi2.InvokeDirect(target: default, arg1: tr);

        // Just like these methods can be called today:
        ChangeIt1(span);
        ChangeIt2(ref span);
    }

    public static void ChangeIt1(Span<int> span) { }
    public static void ChangeIt2(ref Span<int> span) { }
}
```

FWIW `__arglist` (only supported on Windows) currently compiles with byref-like types:
```cs
Span<byte> s1 = stackalloc byte[1];
Span<byte> s2 = stackalloc byte[11];
Span<byte> s3 = stackalloc byte[111];
CallMe(__arglist(s1, s2, s3));

static unsafe void CallMe(__arglist)
{
    // However, when enumerating __arglist here, the runtime throws when accessing a byref-like
    // type although that limitation is easily fixable on Windows (just a runtime limitation; not compiler)
```

# Future
Holding area of features discussed but not planned yet.

## Variable-length, safe collections

> The API proposal below does have a variable-lenth stack-only approach that uses an internal GC tracking mechanism. A easier-to-pass or callback version is not expected in 8.0; see https://github.com/dotnet/runtime/issues/75349.

Reviewer 10: ```suggestion
The API proposal below does have a variable-length stack-only approach that uses an internal GC tracking mechanism. A easier-to-pass or callback version is not expected in 8.0; see https://github.com/dotnet/runtime/issues/75349.
```

- Using `ref byte`. This is possible in 7.0 due to the new "ref field" support. Previously, there was an internal `ByReference<T>` class that was used. This `ref byte` approach is used today in reflection invoke when there are <=4 parameters.

> - Using `IntPtr` with tracking or pinning. Currently, tracking is only supported internally through the use of a "RegisterForGCReporting()" mechanism. This approach is used today in reflection invoke when there are >=5 parameters.

Reviewer 3: ```suggestion
- Using `void*` with tracking or pinning. Currently, tracking is only supported internally through the use of a "RegisterForGCReporting()" mechanism. This approach is used today in reflection invoke when there are >=5 parameters.
```
We are using `void*` for pointers in new APIs.

> This design proposes the use of `TypedReference` for the safe version of APIs, and either `ref byte`, `IntPtr` or `TypeReference*` approaches for the unsafe version. The `ref byte` and `IntPtr` approaches require the use of pointers through the internal `object.GetRawData()` and other means which would need to be exposed publicaly or wrapped in some manner.

Reviewer 3: ```suggestion
This design proposes the use of `TypedReference` for the safe version of APIs, and either `ref byte`, `void*` or `TypeReference*` approaches for the unsafe version. The `ref byte` and `IntPtr` approaches require the use of pointers through the internal `object.GetRawData()` and other means which would need to be exposed publicaly or wrapped in some manner.
```

## TypedReference

> The internal reflection implementation will not likely be based on `TypedReference` -- instead it will take, from public APIs, `TypedReference` instances and "peel off" the interior pointer either as a `ref byte` or as an `IntPtr` along with tracking to support GC. The use of `TypedReference` is essentially to not require the use of unsafe code in the public APIs. Unsafe public APIs will likely be added as well that should have slightly better performance characteristics.

Reviewer 3: ```suggestion
The internal reflection implementation will not likely be based on `TypedReference` -- instead it will take, from public APIs, `TypedReference` instances and "peel off" the interior pointer either as a `ref byte` or as an `void*` along with tracking to support GC. The use of `TypedReference` is essentially to not require the use of unsafe code in the public APIs. Unsafe public APIs will likely be added as well that should have slightly better performance characteristics.
```
